protocol.md

```diff
@@ -75,6 +75,9 @@ The basic premise of Dark Crystal is that the secret owner trusts their custodia
 In addition, using a PKI also means that public/private keys are assigned to (partially) identified individuals.
 The advantage of using Magic Wormhole is that they do not need to disclose this to any other parties.
 
+> - The biggest pain-point for security with Dark Crystal is when returning shards during recovery, we need to be sure we are sending them to the right person. This might be a situation where the secret owner has lost their computer and has limited ways they can confirm their identity. Magic Wormhole gives us an easy-to-use and secure way to do out-of-band authentication.
+> - Is it worth mentioning the possibility that Magic Wormhole could be used to receive shards in the browser without the 'custodians' needing to install anything? I don't know whether some of the protocol details below would make this difficult / impossible. But this would greatly broaden who i can choose as a trusted contact, and give it an advantage over other platforms / protocols.
```

Comment:
Magic Wormhole can be conceived as two "layers": the core mailbox protocol, that allows two computers to establish an encrypted secure channel; and the "file-transfer protocol" that is used to exchange files.

When I wrote this draft, I was imaging this "dark crystal custodian protocol" being a new application-level protocol -- that is, replacing the "file transfer" portion. _However_, if we take out the challenges etc it might indeed to possible to leverage the file-transfer protocol -- and hence existing apps! -- to do this part.

Although we're developing a "next generation" file-transfer protocol with bi-direction and multiple file features, currently a file-transfer is one-time and in one direction. So, this would preclude e.g. the challenge/response identity checks -- but without identity, we don't need that.

So, I think this is well worth considering: that is, applications could shell out to `wormhole send` or direct users to use the Web client (for sending _or_ receiving) so long as we can express the operation as "produce some files" and then "send each of those files to one Custodian".

Interestingly, this kind of mirrors how one might ask friends to store physical secrets -- e.g. shared keys to a physical lock-box or house. That is, "hey neighbour, can you keep a house-key of mine please?" mirrored by "oops, I forgot my housekey .. hey neighbour, can I have it back please?". A one-time operation in each direction.

/cc @sylvia-la

Reply:
i think it would be a shame to need to take out the challenges part.

i was imagining, rather than use the existing web client, that we would make a new one for this protocol.

But to clarify, the browser-based client would be only for the custodians.  I think its fine that the secret-owner needs to install some software locally, but for the custodians, who are just helping out, its seems like a big ask. This means the multiple file sending / receiving feature wouldn't need to work in the browser, and nor does the secret-sharing / recovery bit - as its only the secret owner who needs those.

so i guess the main limitation of this would be what networking stuff we can use for the transit connection. (eg: needing to use something like webRTC instead of TCP/ UDP). If i understood right we would anyway only use a transit connection for 'big' secrets. So a compromise might be you can only use the browser based client for 'small' secret shares.

the house key thing is a good analogy.  we did some role play game for some user testing sessions involving people going on holiday for a long time and deciding what to do with house keys.

Reply:
We've already added WebSockets to the transit relay and are working on a Web-based file-transfer application.

So, I believe all the pieces are available to make a Web application that Custodians could use (with or without a custom application-layer protocol).

The limitation is that _all_ connections would go via the relay. So, this would leak some metadata to the relay (and likely network observers) but of course all the actual communications would still be encrypted. Using WebRTC _could_ be an alternative to this WebSocket relay connection (but we haven't explored that ourselves yet).

Said a different way: the current approach to NAT punching and "how to connect to another web page" is to use the relay (that is, all Web clients connect via WebSocket to the relay). It _would_ be interesting to explore WebRTC as well for this use-case though!

As you suggest, we could forget about all that if the actual Shards are just shared via the mailbox / rendezvous server (i.e. no "direct" connection, via relay or anything else).

Reply:
re: the house-key analogy is any of that available online? It certainly sounds interesting ... but of course with SSS probably gets weird ("how do I share 1/3rd of a house key?"). ;)

Reply:
all makes sense. 

so there bit of a usability / security trade-off, and maybe some users would prefer that the custodians don't use a browser-based client. but we could say in the motivation something like:

'Optionally the custodians can use a browser based client with no installation or configuration beforehand, making it a realistic option for non-technical users, overcoming many of the usability issues with peer-to-peer protocols'

re: house-key thing. it is not written up in much detail, but see page 13 of [this report](https://darkcrystal.pw/assets/briar-case-project-report.pdf)